c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1474,413 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+  }
+
+  // -------------------------------------------------------
+  // Event implementation
+
+  Maybe<Own<Event>> fire() override {
+    // Call Awaiter::await_resume() and proceed with the coroutine. Note that this will not destroy
+    // the coroutine if control flows off the end of it, because we return suspend_always() from
+    // final_suspend().
+    //
+    // It's tempting to arrange to check for exceptions right now and reject the promise that owns
+    // us without resuming the coroutine, which would save us from throwing an exception when we
+    // already know where it's going. But, we don't really know: unlike in the KJ_NO_EXCEPTIONS
+    // case, the `co_await` might be in a try-catch block, so we have no choice but to resume and
+    // throw later.
+    //
+    // TODO(someday): If we ever support coroutines with -fno-exceptions, we'll need to reject the
+    //   enclosing coroutine promise here, if the Awaiter's result is exceptional.
+    auto coroutine = Handle::from_promise(*this);
+    coroutine.resume();
+
+    return nullptr;
+  }
+
+  void fulfill(FixVoid<T>&& value) {
+    // Called by the return_value()/return_void() functions in our base class.
+
+    if (waiting) {
+      result = kj::mv(value);
+      onReadyEvent.arm();
+      waiting = false;
+    }
+  }
+
+  OnReadyEvent onReadyEvent;
+  ExceptionOr<FixVoid<T>> result;
+  bool waiting = true;
+
+  UnwindDetector unwindDetector;
+
+  struct DisposalResults {
+    bool destructorRan = false;
+    Maybe<Exception> exception;
+  };
+  Maybe<DisposalResults&> maybeDisposalResults;
+  // Only non-null during destruction. Before calling coroutine.destroy(), our disposer sets this
+  // to point to a DisposalResults on the stack so unhandled_exception() will have some place to
+  // store unwind exceptions. We can't store them in this Coroutine, because we'll be destroyed once
+  // coroutine.destroy() has returned. Our disposer then rethrows as needed.
+
+  friend class CoroutineBase<Coroutine<T>, T>;
+};
+
+template <typename Self, typename T>
+class CoroutineBase {
+public:
+  void return_value(T value) {
+    static_cast<Self*>(this)->fulfill(kj::mv(value));
+  }
+};
+template <typename Self>
+class CoroutineBase<Self, void> {
+public:
+  void return_void() {
+    static_cast<Self*>(this)->fulfill(_::Void());
+  }
+};
+// The Coroutines spec has no `_::FixVoid<T>` equivalent to unify valueful and valueless co_return
+// statements, and programs are ill-formed if the coroutine implementation object (Coroutine<T>) has
+// both a `return_value()` and `return_void()`. No amount of EnableIffery can get around it, so
+// these return_* functions live in a CRTP mixin.
+
+static stdcoro::coroutine_handle<void> noopCoroutine() noexcept;
+// TODO(msvc): Coroutines TS implementation doesn't have std::experimental::noop_coroutine(), so
+//   we'll roll our own below. This function is static so we can keep libkj coroutines header-only.
```

Comment:
Hmm, why do we want it to be header-only?

Reply:
My thinking here was that it'd be nice if a dependent project can use coroutines without worrying about whether their capnproto package was built with or without `-fcoroutines-ts`. But I think your comment about `await_suspend()` returning bool might obviate the concern (for now).